core/src/main/java/org/apache/calcite/rel/rules/AggregateReduceFunctionsRule.java

```diff
@@ -97,12 +98,96 @@
+      this.reduceAvg = true;
+      this.reduceSum = true;
+      this.reduceStddevPop = true;
+      this.reduceStddevSamp = true;
+      this.reduceVarPop = true;
+      this.reduceVarSamp = true;
+      this.reduceCovarPop = true;
+      this.reduceCovarSamp = true;
+      this.reduceRegrSxx = true;
+      this.reduceRegrSyy = true;
+    }
+
+    public AggregateFunctionsToReduce(final boolean reduceAvg, final boolean reduceSum,
+        final boolean reduceStddevPop, final boolean reduceStddevSamp,
+        final boolean reduceVarPop, final boolean reduceVarSamp,
+        final boolean reduceCovarPop, final boolean reduceCovarSamp,
+        final boolean reduceRegrSxx, final boolean reduceRegrSyy) {
+      this.reduceAvg = reduceAvg;
+      this.reduceSum = reduceSum;
+      this.reduceStddevPop = reduceStddevPop;
+      this.reduceStddevSamp = reduceStddevSamp;
+      this.reduceVarPop = reduceVarPop;
+      this.reduceVarSamp = reduceVarSamp;
+      this.reduceCovarPop = reduceCovarPop;
+      this.reduceCovarSamp = reduceCovarSamp;
+      this.reduceRegrSxx = reduceRegrSxx;
+      this.reduceRegrSyy = reduceRegrSyy;
+    }
+  }
+
+  /**
+   * Gets an instance of AggregateReduceFunctionsRule
+   * with client provided specific functions to reduce
+   * @param functionsToReduce client provided information
+   *                          on which specific functions will be
+   *                          reduced by this rule
+   * @return an instance of AggregateReduceFunctionsRule
+   */
+  public static AggregateReduceFunctionsRule
+      getInstanceWithSpecificFunctionsToReduce(final AggregateFunctionsToReduce functionsToReduce) {
+    Preconditions.checkArgument(functionsToReduce != null,
+          "Error: expecting a valid handle for AggregateFunctionsToReduce");
+    return new AggregateReduceFunctionsRule(operand(LogicalAggregate.class, any()),
+      RelFactories.LOGICAL_BUILDER, functionsToReduce);
+  }
+
   //~ Constructors -----------------------------------------------------------
 
   /** Creates an AggregateReduceFunctionsRule. */
   public AggregateReduceFunctionsRule(RelOptRuleOperand operand,
       RelBuilderFactory relBuilderFactory) {
     super(operand, relBuilderFactory, null);
+    // by default, this rule will reduce all functions it handles
+    this.functionsToReduce = new AggregateFunctionsToReduce();
+  }
+
+  /**
+   * Creates an AggregateReduceFunctionsRule with client
+   * provided information on which specific functions will
+   * be reduced by this rule
+   * @param operand root operand
+   * @param relBuilderFactory builder for relational expressions
+   * @param functionsToReduce client provided information
+   *                          on which specific functions
+   *                          will be reduced by this rule
+   */
+  public AggregateReduceFunctionsRule(RelOptRuleOperand operand,
+      RelBuilderFactory relBuilderFactory, final AggregateFunctionsToReduce functionsToReduce) {
```

Comment:
I see you created a new object for holding the information of which functions to reduce. Wouldn't be simpler to just pass a `Set<SqlKind> functionsToReduce`?

Reply:
Thanks. I am using a Set<SqlKind> now.

Reply:
Great, thanks!